client/components/data/query-terms/index.jsx

```diff
@@ -1,68 +1,30 @@
-import isShallowEqual from '@wordpress/is-shallow-equal';
 import PropTypes from 'prop-types';
-import { Component } from 'react';
-import { connect } from 'react-redux';
+import { useEffect } from 'react';
+import { useDispatch } from 'react-redux';
 import { requestSiteTerms } from 'calypso/state/terms/actions';
 import { isRequestingTermsForQuery } from 'calypso/state/terms/selectors';
 
-class QueryTerms extends Component {
-	UNSAFE_componentWillMount() {
-		this.request( this.props );
-	}
-
-	UNSAFE_componentWillReceiveProps( nextProps ) {
-		if (
-			this.props.siteId === nextProps.siteId &&
-			this.props.taxonomy === nextProps.taxonomy &&
-			isShallowEqual( this.props.query, nextProps.query )
-		) {
-			return;
-		}
-
-		this.request( nextProps );
+const request = ( siteId, taxonomy, query ) => ( dispatch, getState ) => {
+	if ( siteId && ! isRequestingTermsForQuery( getState(), siteId, taxonomy, query ) ) {
+		dispatch( requestSiteTerms( siteId, taxonomy, query ) );
 	}
+};
 
-	request( props ) {
-		if ( props.requesting || ! props.siteId ) {
-			return;
-		}
+function QueryTerms( { siteId, taxonomy, query = {} } ) {
+	const dispatch = useDispatch();
+	const stringifiedQuery = JSON.stringify( query );
```

Comment:
Stringifying and parsing it again works, and I even found it recommended in a Dan Abramov tweet, but the following memoization hook seems more elegant to me:
```js
function useMemoized( obj ) {
  const memoObj = useRef();
  if ( ! isShallowEqual( obj, memoObj.current ) ) {
    memoObj.current = obj;
  }
  return memoObj;
}
```
It updates the `memoObj` object only when the new one is shallowly different from the old one, otherwise it ignores the `obj` param and returns the old value.

Reply:
@jsnajdr thanks, I added https://github.com/Automattic/wp-calypso/pull/56572/commits/c9f802b4eb0acaf9038eaa05b061e6cafea8c344 – let me know what you think. Tbh I think the previous solution is  simpler but using `isShallowEqual` is probably more safe to use.

Reply:
The thing I like about `useMemoized` is that abstracts the memoization details away, and that it can potentially take a `comparator` argument, making it very universal.